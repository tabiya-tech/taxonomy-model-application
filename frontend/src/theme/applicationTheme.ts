import { createTheme, PaletteOptions, ThemeOptions, SimplePaletteColorOptions } from "@mui/material/styles/";
import { CSSClampFnCalculatorPx, CSSClampFnCalculatorRem, ScreenSize } from "src/utils/CSSClampFnCalculator";

export enum ThemeMode {
  LIGHT = "light",
  DARK = "dark",
}

// @ts-ignore
const _temp_palette = createTheme({
  palette: {
    contrastThreshold: 4.5, //WCAG 2.0 (AA) ensure color-contrast is at least 4.5:1
  },
}).palette;

const augmentedThemeColor = (color: string, contrastColor?: string) =>
  _temp_palette.augmentColor({
    color: {
      main: color,
      contrastText: contrastColor,
    },
  });

const TabiyaBasicColors = {
  DarkBlue: "#002147",
  LightBlue: "#265EA7",
  Yellow: "#EEFF41",
  Green: "#00FF91",
  Gray: "#F3F1EE",
  GrayDark: "#43474E",
};
export const TabiyaIconStyles = {
  fontSizeSmall: {
    fontSize: "1rem",
  },
  fontSizeMedium: {
    fontSize: "1.5rem",
  },
  fontSizeLarge: {
    fontSize: "2.5rem",
  },
  root: {
    fontSize: "1.5rem",
  },
};

const lightPalette: PaletteOptions = {
  contrastThreshold: 4.5, // WCAG 2.0 (AA) ensure color-contrast is at least 4.5:1
  primary: augmentedThemeColor(TabiyaBasicColors.DarkBlue),
  secondary: augmentedThemeColor(TabiyaBasicColors.LightBlue),
  tabiyaYellow: augmentedThemeColor(TabiyaBasicColors.Yellow),
  tabiyaGreen: augmentedThemeColor(TabiyaBasicColors.Green),
  containerBackground: {
    light: "#FFFFFF",
    dark: "#DFDDD9",
    main: TabiyaBasicColors.Gray,
    contrastText: TabiyaBasicColors.GrayDark,
  },
  error: { ...augmentedThemeColor("#FF5449"), dark: "#690005", light: "#FFEDEA" },
  warning: { ...augmentedThemeColor("#FDAB40", TabiyaBasicColors.GrayDark), dark: "#B84204", light: "#FFF3E0" },
  info: { ...augmentedThemeColor("#4FC3F7"), dark: "#003662", light: "#CAF5FF" },
  success: { ...augmentedThemeColor("#6BF0AE"), dark: "#1D6023", light: "#E8F5E9" },
  grey: {
    900: "#211F1D",
    800: "#41403D",
    700: "#605E5B",
    600: "#74726F",
    500: "#9D9B98",
    400: "#BBB9B5",
    300: "#DFDDD9",
    200: "#EDEBE8",
    100: TabiyaBasicColors.Gray,
    50: "#F8F6F3",
    A100: "#F0F3EE",
    A200: "#D9F5D3",
    A400: "#C6F7A4",
    A700: "#CAF09D",
  },
  text: {
    primary: TabiyaBasicColors.DarkBlue,
    secondary: TabiyaBasicColors.GrayDark,
    textAccent: TabiyaBasicColors.LightBlue,
    textWhite: "#FFFFFF",
    textBlack: "#000000",
    disabled: "#E0E2EC",
  },
  common: {
    white: "#ffffff",
    black: "#000000",
  },
};

const darkPalette: PaletteOptions = {
  // Add Some dark theme palette options,
  // currently using the light theme palette
  ...lightPalette,
};

export const TabiyaBaseSizes = {
  // All sizes are in px
  spacing: 8,
  rounding: 8,
  font: 16,
};
export const applicationTheme = (theme: ThemeMode) => {
  const screenSizePx: ScreenSize = {
    minWidth: 800,
    maxWidth: 1200,
    minHeight: 500,
    maxHeight: 800,
  };
  const screenSizeRem: ScreenSize = {
    minWidth: screenSizePx.minWidth / TabiyaBaseSizes.font,
    maxWidth: screenSizePx.maxWidth / TabiyaBaseSizes.font,
    minHeight: screenSizePx.minHeight / TabiyaBaseSizes.font,
    maxHeight: screenSizePx.maxHeight / TabiyaBaseSizes.font,
  };

  // cache the clamp functions to avoid recalculating them
  const spacingClampFn = CSSClampFnCalculatorPx(TabiyaBaseSizes.spacing / 4, TabiyaBaseSizes.spacing, screenSizePx);
  const roundingClampFn = CSSClampFnCalculatorPx(TabiyaBaseSizes.rounding / 4, TabiyaBaseSizes.rounding, screenSizePx);

  const activePalette: PaletteOptions = theme === ThemeMode.LIGHT ? lightPalette : darkPalette;
  const activeTheme: ThemeOptions = {
    palette: activePalette,
    spacing: (factor: number) => `calc(${spacingClampFn} * ${factor})`,
    fixedSpacing: (factor: number) => `${factor * TabiyaBaseSizes.spacing}px`,
    responsiveBorderRounding: (factor: number | "50%") => {
      if (factor === "50%") {
        // percentage makes no sense to be responsive
        return factor;
      }
      return `calc(${roundingClampFn} * ${factor})`;
    },
    rounding: (factor: number | "50%") => {
      if (factor === "50%") {
        // percentage makes no sense to multiply with factor
        return factor;
      }
      return `${factor * TabiyaBaseSizes.rounding}px`;
    },
    tabiyaSpacing: {
      none: 0,
      xs: 0.5,
      sm: 1,
      md: 2,
      lg: 3,
      xl: 4,
    },
    shape: {
      borderRadius: 8,
    },
    tabiyaRounding: {
      none: 0,
      xs: 0.5,
      sm: 1,
      md: 2,
      lg: 3,
      xl: 4,
      full: "50%",
    },
    typography: {
      htmlFontSize: TabiyaBaseSizes.font, // Set the base font size
      fontFamily: "Inter, sans-serif", // Set the desired font family
      // @ts-ignore
      fontSize: TabiyaBaseSizes.font, // Set the base font size
      fontWeightLight: 300,
      fontWeightRegular: 400,
      fontWeightMedium: 500,
      fontWeightBold: 700,
      h1: {
        fontFamily: "IBM Plex Mono",
        fontWeight: "700",
        fontSize: CSSClampFnCalculatorRem(1.3, 2, screenSizeRem),
        color: activePalette.text!!.primary,
      },
      h2: {
        fontFamily: "IBM Plex Mono",
        fontWeight: "700",
        fontSize: CSSClampFnCalculatorRem(1.25, 1.85, screenSizeRem),
        color: activePalette.text!!.primary,
      },
      h3: {
        fontFamily: "IBM Plex Mono",
        fontWeight: "700",
        fontSize: CSSClampFnCalculatorRem(1.2, 1.7, screenSizeRem),
        color: activePalette.text!!.primary,
      },
      h4: {
        fontFamily: "IBM Plex Mono",
        fontWeight: "700",
        fontSize: CSSClampFnCalculatorRem(1.15, 1.55, screenSizeRem),
        color: activePalette.text!!.primary,
      },
      h5: {
        fontFamily: "IBM Plex Mono",
        fontWeight: "700",
        fontSize: CSSClampFnCalculatorRem(1.1, 1.4, screenSizeRem),
        color: activePalette.text!!.primary,
      },
      h6: {
        fontFamily: "IBM Plex Mono",
        fontWeight: "700",
        fontSize: CSSClampFnCalculatorRem(1, 1.25, screenSizeRem),
        color: activePalette.text!!.primary,
      },
      subtitle1: {
        fontFamily: "Inter",
        fontWeight: "500",
        fontSize: CSSClampFnCalculatorRem(0.75, 1, screenSizeRem),
        color: activePalette.text!!.textAccent,
      },
      subtitle2: {
        fontFamily: "Inter",
        fontWeight: "500",
        fontSize: CSSClampFnCalculatorRem(0.5, 0.875, screenSizeRem),
        color: activePalette.text!!.textAccent,
      },
      body1: {
        fontFamily: "Inter",
        fontWeight: "400",
        fontSize: CSSClampFnCalculatorRem(0.75, 1, screenSizeRem),
        color: activePalette.text!!.secondary,
      },
      body2: {
        fontFamily: "Inter",
        fontWeight: "400",
        fontSize: CSSClampFnCalculatorRem(0.5, 0.875, screenSizeRem),
        color: activePalette.text!!.secondary,
      },
      button: {
        fontFamily: "Inter",
        fontWeight: "500",
        fontSize: CSSClampFnCalculatorRem(0.75, 1, screenSizeRem),
        color: activePalette.text!!.primary,
        textTransform: "none",
      },
      caption: {
        fontFamily: "Inter",
        fontWeight: "400",
        fontSize: CSSClampFnCalculatorRem(0.5, 0.75, screenSizeRem),
      },
      overline: {
        fontFamily: "Inter",
        fontWeight: "400",
        fontSize: CSSClampFnCalculatorRem(0.5, 0.75, screenSizeRem),
      },
    },
    components: {
      MuiDialogTitle: {
        defaultProps: {
          variant: "h2",
        },
      },
      MuiFormLabel: {
        styleOverrides: {
          asterisk: {
            color: (activePalette.error as SimplePaletteColorOptions).main, // unclear why typescript complaints about this and we need to cast
          },
        },
      },
      MuiInput: {
        styleOverrides: {
          input: {
            "::placeholder": {
              color: activePalette.text!!.secondary,
            },
          },
        },
      },
      MuiTableHead: {
        defaultProps: {
          style: {
            background: (activePalette.containerBackground as SimplePaletteColorOptions)!!.main,
          },
        },
      },
      MuiSvgIcon: {
        styleOverrides: {
          ...TabiyaIconStyles,
        },
      },
      MuiIcon: {
        styleOverrides: {
          ...TabiyaIconStyles,
        },
      },
    },
  };
  return createTheme(activeTheme);
};

export default applicationTheme;
