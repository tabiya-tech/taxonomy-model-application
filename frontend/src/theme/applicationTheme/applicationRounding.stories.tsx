import { Meta, StoryObj } from "@storybook/react";
import { RoundingElements } from "src/_test_utilities/SpacingAndRoundingElements";

const meta: Meta = {
  title: "Style/Rounding",
  tags: ["autodocs"],
  component: RoundingElements,
  argTypes: {},
};

export default meta;

type Story = StoryObj;

export const RoundingStyles: Story = {};
