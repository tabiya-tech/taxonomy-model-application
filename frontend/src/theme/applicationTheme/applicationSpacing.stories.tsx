import { Meta, StoryObj } from "@storybook/react";
import { SpacingElements } from "src/_test_utilities/SpacingAndRoundingElements";

const meta: Meta = {
  title: "Style/Spacing",
  component: SpacingElements,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;

type Story = StoryObj;

export const SpacingStyles: Story = {
  args: {
    children: <SpacingElements />,
  },
};
