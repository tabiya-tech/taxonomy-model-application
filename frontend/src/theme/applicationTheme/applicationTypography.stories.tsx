import { Meta, StoryObj } from "@storybook/react";
import { Box, Theme, Typography, useTheme } from "@mui/material";
import { useRef } from "react";
import { useResponsiveStyleValue } from "src/theme/applicationTheme/useResponsiveStyleValue";
import { TabiyaBaseSizes } from "src/theme/applicationTheme/applicationTheme";

const meta: Meta = {
  title: "Style/Typography",
  component: Box,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;

type Story = StoryObj;

interface TypographyElementProps {
  theme: Theme;
  typographyKey: string;
}
const TypographyElement = (props: TypographyElementProps) => {
  const typographyRef = useRef<HTMLDivElement>(null);
  const computedFontSizeInPx = useResponsiveStyleValue(typographyRef, "font-size");
  const typographyObjectKey = props.typographyKey as keyof typeof props.theme.typography;
  const typographyVariant = props.theme.typography[typographyObjectKey];
  const computedFontSizeInRem = computedFontSizeInPx
    ? Math.round((parseFloat(computedFontSizeInPx) / TabiyaBaseSizes.font) * 100) / 100
    : "";
  if (
    typeof typographyVariant === "object" &&
    typographyVariant !== null &&
    "fontFamily" in typographyVariant &&
    "fontSize" in typographyVariant &&
    props.typographyKey !== "inherit"
  ) {
    return (
      <Typography variant={props.typographyKey as any} ref={typographyRef}>
        {`<${props.typographyKey}> ${typographyVariant.fontFamily} ${computedFontSizeInPx} = ${computedFontSizeInRem}rem`}
      </Typography>
    );
  }
  return null;
};
const TypographyElements = () => {
  const theme = useTheme();
  return (
    <Box
      sx={{
        display: "flex",
        flexDirection: "column",
        padding: theme.tabiyaSpacing.lg,
        gap: theme.tabiyaSpacing.lg,
      }}
    >
      <Typography variant={"h4"}>Basic Typography</Typography>
      <Box
        sx={{
          gap: theme.tabiyaSpacing.md,
          display: "flex",
          flexDirection: "column",
          alignItems: "start",
        }}
      >
        {Object.keys(theme.typography as Record<string, any>).map((typographyKey: string) => (
          <TypographyElement key={typographyKey} typographyKey={typographyKey} theme={theme} />
        ))}
      </Box>
      <Typography variant={"h4"}>Font Weight</Typography>
      <Box
        sx={{
          gap: theme.tabiyaSpacing.md,
          display: "flex",
          flexDirection: "column",
          alignItems: "start",
        }}
      >
        <Typography sx={{ fontWeight: theme.typography.fontWeightBold }}>
          {`fontWeightBold: ${theme.typography.fontWeightBold}`}
        </Typography>
        <Typography sx={{ fontWeight: theme.typography.fontWeightMedium }}>
          {`fontWeightMedium: ${theme.typography.fontWeightMedium}`}
        </Typography>
        <Typography sx={{ fontWeight: theme.typography.fontWeightRegular }}>
          {`fontWeightRegular: ${theme.typography.fontWeightRegular}`}
        </Typography>
        <Typography sx={{ fontWeight: theme.typography.fontWeightLight }}>
          {`fontWeightLight: ${theme.typography.fontWeightLight}`}
        </Typography>
      </Box>
    </Box>
  );
};

export const TypographyStyles: Story = {
  args: {
    children: <TypographyElements />,
  },
};
