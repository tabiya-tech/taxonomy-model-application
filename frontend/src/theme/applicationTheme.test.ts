import applicationTheme, { ThemeMode } from "./applicationTheme";

describe("applicationTheme", () => {
  // should return the Light theme
  it("should return the Light theme", () => {
    applicationTheme(ThemeMode.LIGHT);
    expect(applicationTheme).toMatchSnapshot();
  });
  // should return the Dark theme
  it("should return the Dark theme", () => {
    applicationTheme(ThemeMode.DARK);
    expect(applicationTheme).toMatchSnapshot();
  });
});
