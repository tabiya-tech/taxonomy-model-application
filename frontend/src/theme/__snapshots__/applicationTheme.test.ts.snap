// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applicationTheme should return the Dark theme 1`] = `[Function]`;

exports[`applicationTheme should return the Light theme 1`] = `[Function]`;
